docs(ci): Improve CI profiling docs (#153)
pyroscope-io/docs: docs/ci.mdx

import TabItem from '@theme/TabItem';
import Link from '@docusaurus/Link';

# Running Pyroscope in a CI/CD pipeline

If you're running tests in a CI/CD pipeline, you can use Pyroscope to profile your tests and upload the results to Pyroscope Cloud.
This allows you to identify performance bottlenecks and optimize your code.

> To use Pyroscope in a CI/CD pipeline, you'll need to install the pyroscope-ci binary. This binary provides the functionality for

Reviewer: ```suggestion
To use Pyroscope in a CI/CD pipeline, you'll need to install the `pyroscope-ci` binary. This binary provides the functionality for
```

profiling tests and uploading the results.

> 

Reviewer: Right here would be a good place for a screenshot of the cloud CI.

> ## Installation

Reviewer: this section is good

The `pyroscope-ci` binary can be downloaded from the [Github Releases page](https://github.com/pyroscope-io/ci/releases).
A convenient installation script is available to automate this process:
```bash
curl https://raw.githubusercontent.com/pyroscope-io/ci/main/install.sh | sh
```
You can also download the script and inspect its contents before running it. We recommend pinning the installation to a fixed commit.

> After running the installation script, you'll have access to the pyroscope-ci binary in your project. You can run it using:

Reviewer: ```suggestion
After running the installation script, you'll have access to the `pyroscope-ci` binary in your project. You can run it using:
```

```
./pyroscope-ci [COMMAND]
```

If you're using Github Actions, we recommend using [`jaxxstorm/action-install-gh-release`](https://github.com/jaxxstorm/action-install-gh-release) action to install the `pyroscope-ci` binary.
For a more complete example see the [Integrating with Github Actions](#integrating-with-github-actions) section below.

## CI profiling steps
To use Pyroscope in your CI/CD pipeline, you'll need to perform three steps: setup, execution, and upload.

1. **Setup:** In this step, you'll initialize Pyroscope and start profiling. This is typically done in a setup script that runs before your tests.
2. **Execution:** In this step, you'll run your tests while Pyroscope is profiling them. Pyroscope will record performance data for each test.

> 3. **Upload/export:** In this step, you'll upload or export the profiling data. This allows you to view the data and analyze the performance of your code.

Reviewer: At first I was really happy that you split this into 3 steps.

But then I kept reading and I no longer think it's a good idea. These 3 sections don't correspond to the sections in the docs. Setup section is called "Profiling go/node tests", upload/export is really more like a subsection of Execution.

It's also a bit unclear as to why Installation is not one of these steps.

Reviewer: It's also a bit confusing that we have Install and Setup

    {label: 'Nodejs (jest)', value: 'nodejs-jest'},
  ]}>
  <TabItem value="go">

>   ## Profiling Go tests

Reviewer: needs a new line to show up

```suggestion

    ## Profiling Go tests
```

  To profile your Go tests with Pyroscope, you can use the `./pyroscope-ci go` command.
  This command will recursively search for all Go test packages under the `pkg` directory, generate a `pyroscope_test.go` file for each package,
  and instrument the test functions with the Pyroscope agent.

  Here's an example of how to use pyroscope-ci go to profile your tests:
  ```bash
  # Note you may need to pull the `pyroscope` dependency, if you are not using it yet.
  # go mod tidy
  ./pyroscope-ci go install --applicationName=myapp pkg
  ```

>   Note that the --applicationName flag specifies the name of the application being profiled.

Reviewer: ```suggestion
  Note that the `--applicationName` flag specifies the name of the application being profiled.
```

>   We recommend running as a ci step, so that you don't have to commit these lines and profile tests when running locally.

Reviewer: ```suggestion
  We recommend running as a CI step, so that you don't have to commit these lines and profile tests when running locally.
```

  That's it! Pyroscope will now profile your tests in your CI/CD pipeline.

  </TabItem>
  <TabItem value="nodejs-jest">

>   ## Profiling Node.js tests

Reviewer: What do you guys think about calling this "Profiling Java Script Tests"? @eh-am @Rperry2174 

I feel like that's slightly better because it's not just node projects that could benefit from this, it's frontend too.

Reviewer 2: I kinda like this for seo purposes but don't feel strongly either way

  To use Pyroscope with Node.js, you'll need to install the @pyroscope/nodejs agent. You can do this via npm or yarn:

  ```bash
  npm install --save-dev @pyroscope/nodejs
  # or
  yarn add --dev @pyroscope/nodejs
  ```

>   ### Setting up Pyroscope in node

Reviewer: ```suggestion
  ### Setting up Pyroscope in Node
```

  To set up Pyroscope, you'll need to create two new files in your project (or edit them if they already exist):

  1. `globalSetup.js`: This file initializes Pyroscope and starts profiling. 
  Pyroscope will only profile your code when the CI environment variable is turned on. Here's an example:
  ```js
  const Pyroscope = require('@pyroscope/nodejs');

  module.exports = () => {
    Pyroscope.init({
      serverAddress: '_',
      appName: 'myapp.frontend',
    });
    if (process.env.CI) {
      Pyroscope.start();
      // Pyroscope.startCpuProfiling(); // Use this for better performance if heap profiling isn't needed
    }
  };
  ```

>   If you don't care about heap profiling, you can replace .start() with .startCpuProfiling() to reduce the performance overhead.

Reviewer: ```suggestion
  If you don't care about heap profiling, you can replace `.start()` with `.startCpuProfiling()` to reduce the performance overhead.
```

  2. `globalTeardown.js`: This file stops profiling and exports the profiling data. Here's an example:
  ```js
    const Pyroscope = require('@pyroscope/nodejs');
    module.exports = (config) => {
      if (process.env.CI) { // Note: be sure to set this ENV variable
        Pyroscope.stopCpuProfiling();
        Pyroscope.stopHeapProfiling();
      }
    };
  ```
  Notice how we conditionally start Pyroscope when the `CI` environment variable is set to `true` so that by default we
  don't profile tests when running locally.

  ### Configuring Jest
  To tell Jest to use these files for configuration, add the following to your `jest.config.js` file:

  ```js
  module.exports = {
    // ...
    globalSetup: '<rootDir>/globalSetup.js',
    globalTeardown: '<rootDir>/globalTeardown.js',
    // ...
  }
  ```
  That's it! Pyroscope will now profile your tests in your CI/CD pipeline.
  
  </TabItem>
</Tabs>

## Execution
After your tests have been set up, it's time to execute them:

<Tabs
  defaultValue="go"
  values={[
    {label: 'Go', value: 'go'},
    {label: 'Nodejs (jest)', value: 'nodejs-jest'},
  ]}>
  <TabItem value="go">

>   ## Executing test profiling with `Go` 

Reviewer: ```suggestion
  ## Executing test profiling with Go
```

  In Go, you can use the following command to run tests with Pyroscope:

  ### Export test results locally
  ```
  CI=true ./pyroscope-ci exec --noUpload --export -- go test ./...
  # 
  ```
  This will create profiles in the `--outputDir` (defaults to `./pyroscope-ci`).
  These profiles can be uploaded to [www.flamegraph.com](https://www.flamegraph.com),
  via the [UI](https://www.flamegraph.com), the [API](https://flamegraph.com/openapi) or using the
  [flamegraph.com Github Action](https://github.com/pyroscope-io/flamegraph.com-github-action).

  ### Upload test results to Pyroscope cloud
  ```bash
  ./pyroscope-ci exec --apiKey="MY_API_KEY" -- go test ./...
  ```

  :::note
  Some CI systems will skip certain tests if the files have not been touched.
  To avoid this, you can run your tests without cache using the `-count=1` flag.
  :::

  </TabItem>
  <TabItem value="nodejs-jest">

>   ## Executing test profiling with `jest` 

Reviewer: ```suggestion
  ## Executing test profiling with jest
```

  After your tests have been set up, it's time to execute them. Here's how to run your tests with `jest`:

>   ### Export test results locally

Reviewer: I'd drop this whole section. Right now it reads like I need to run both of these commands (both export and upload)

Reviewer 2: I agree that we should make it clear to do one or the other, but I think we need to keep this because I think we drastically decrease the chances of someone running this if we _only_ show the cloud version... For example when I tried this out I just wanted to see a result before deciding whether or not to send it to Pyroscope cloud

  These profiles can be uploaded to [www.flamegraph.com](https://www.flamegraph.com),
  via the [UI](https://www.flamegraph.com), the [API](https://flamegraph.com/openapi) or using the
  [flamegraph.com Github Action](https://github.com/pyroscope-io/flamegraph.com-github-action).

  ### Upload test results to Pyroscope cloud
  ```bash
  # npm
  pyroscope-ci exec --apiKey="MY_API_KEY" -- npm run test --runInBand

  # or yarn
  pyroscope-ci exec --apiKey="MY_API_KEY" -- yarn run test --runInBand
  ```

  Make sure to include the --runInBand flag so that the profiling session started in globalSetup.js is available to all tests.
  Otherwise, jest will spawn different workers without the ability to perform individual setup/teardown.

  :::note
  Some CI systems will skip certain tests if the files have not been touched.
  To avoid this behavior, you can try running your tests without cache using the --no-cache flag.
  :::

  For more eamples, see the <Link to="https://github.com/pyroscope-io/ci/tree/main/examples/nodejs/jest">example</Link>.
  </TabItem>
</Tabs>

Notice that any command can be set (`make test`, `./test.sh` etc).

> ## Uploading test profiles to Pyroscope Cloud

Reviewer: Not sure about this section. Right now the title reads like I have to do extra actions to upload test profiles. But really this section talks about some details about how profiles are merged and all of that, which I don't think people have any context for at this point. I would maybe drop this section and instead explain these things in a video at the top.

  </TabItem>
  <TabItem value="nodejs-jest">

  ## Node.js

>   To use Pyroscope with Node.js, you'll need to install the @pyroscope/nodejs agent. You can do this via npm or yarn:

Author: ```suggestion
  To use Pyroscope with Node.js, you'll need to install the `@pyroscope/nodejs` agent. You can do this via npm or yarn:
```

> ## Visualizing profiles

Reviewer: I would get rid of this section and put a screenshot of pyroscope cloud CI at the top.

The profiles can be visualized in the [CI page under Pyroscope Cloud](https://pyroscope.cloud/ci).

<iframe frameBorder="0" width="100%" height="400" src="https://flamegraph.com/share/68b22ba6-9d82-11ed-a9bc-e2a53ffd5c62/iframe?colorMode=light&onlyDisplay=flamegraph"></iframe>

For illustration, above is the profile for `github.com/pyroscope-io/pyroscope` go tests,

> which shows that the race condition detector (`racecall`) spends a non-negligible amount of time.

Reviewer: ```suggestion
which shows that the race condition detector (`racecall`) spends a significant amount of time.
```

> # TEMPORARILY MOVING HERE...but considering deleting altogether

Reviewer: yeah I support this
